File: backend/api/v1/notifications.py
"""/api/v1/notifications/prefs — workspace notification preferences.

The founder's Settings -> Notifications surface. v1 stores the PREFERENCES
only (an events x channels enable matrix + a quiet-hours window); the actual
email / Slack delivery wiring is a later phase.

* ``GET  /api/v1/notifications/prefs`` — get-or-create for the active
  workspace. A workspace with no row yet reads the sensible defaults
  (:data:`backend.notifications.db.DEFAULT_MATRIX`), which are then persisted,
  so a later PUT updates a real row.
* ``PUT  /api/v1/notifications/prefs`` — replace the matrix + quiet hours
  wholesale. The matrix is validated against the known event/channel ids and
  quiet-hours times must be ``HH:MM``.

Workspace resolution mirrors :mod:`backend.api.v1.runs` (the ``get_workspace_id``
dep publishes the workspace into the ORM-scoping contextvar). Per-product
overrides from the design are intentionally OMITTED in v1 — this is the global
matrix + quiet hours only.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from backend.api.deps import get_db_session, get_workspace_id
from backend.notifications.bindings import available_channels
from backend.notifications.serialization import (
    PrefsBody,
    PrefsView,
    get_or_create_prefs,
    prefs_view_from_row,
)

router = APIRouter()

_to_view = prefs_view_from_row


_get_or_create = get_or_create_prefs


@router.get("/prefs")
async def get_prefs(
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> PrefsView:
    """Get-or-create the active workspace's notification preferences."""
    row = await _get_or_create(session, workspace_id)
    channels = await available_channels(session, workspace_id=workspace_id)
    return _to_view(row, channels)


@router.put("/prefs")
async def put_prefs(
    payload: PrefsBody,
    workspace_id: Annotated[uuid.UUID, Depends(get_workspace_id)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
) -> PrefsView:
    """Replace the matrix + quiet hours wholesale (single row per workspace)."""
    row = await _get_or_create(session, workspace_id)
    row.matrix = payload.matrix
    row.quiet_hours_enabled = payload.quiet_hours_enabled
    row.quiet_hours_start = payload.quiet_hours_start
    row.quiet_hours_end = payload.quiet_hours_end
    await session.commit()
    await session.refresh(row)
    channels = await available_channels(session, workspace_id=workspace_id)
    return _to_view(row, channels)


__all__ = ["PrefsBody", "PrefsView", "get_prefs", "put_prefs", "router"]

File: apps/pwa/lib/api/notifications.ts
/** Notifications API — REAL backend `/api/v1/notifications/prefs`
 *  (backend/api/v1/notifications.py): the founder's notification preferences,
 *  an events × channels enable matrix plus a quiet-hours window.
 *
 *   GET  /api/v1/notifications/prefs  — get-or-create the prefs for the active
 *                                       workspace (a fresh workspace reads
 *                                       sensible defaults, then persists them)
 *   PUT  /api/v1/notifications/prefs  — replace the matrix + quiet hours
 *                                       wholesale
 *
 *  Both return the `PrefsView` shape ({@link NotificationPrefsView}): the stored
 *  prefs PLUS `available_channels` (the workspace's live notify channels, derived
 *  per read from connector bindings). The PUT BODY is the narrower `PrefsBody`
 *  ({@link NotificationPrefs}) — matrix + quiet hours ONLY. `available_channels`
 *  is response-only, so `updateNotificationPrefs` sends just the body fields (the
 *  backend `extra=forbid` rejects an echoed `available_channels`). */

import { apiFetch } from "./client";
import type { NotificationPrefs, NotificationPrefsView } from "./types";

/** Notification preferences for the active workspace (get-or-create). */
export function getNotificationPrefs(): Promise<NotificationPrefsView> {
  return apiFetch<NotificationPrefsView>("/api/v1/notifications/prefs");
}

/** Replace the matrix + quiet hours wholesale. Sends only the writable
 *  `PrefsBody` fields (never `available_channels`) and returns the persisted
 *  prefs with the freshly-derived channels. */
export function updateNotificationPrefs(prefs: NotificationPrefs): Promise<NotificationPrefsView> {
  const body: NotificationPrefs = {
    matrix: prefs.matrix,
    quiet_hours_enabled: prefs.quiet_hours_enabled,
    quiet_hours_start: prefs.quiet_hours_start,
    quiet_hours_end: prefs.quiet_hours_end,
  };
  return apiFetch<NotificationPrefsView>("/api/v1/notifications/prefs", {
    method: "PUT",
    body: JSON.stringify(body),
  });
}
